### PR TITLE
[Core] Launching cluster autodowned or manually terminated

### LINKED ISSUE #2129: [Python API] NoneType issue when re-launching a cluster terminated manually on the console
An issue from our user, where our python API is not reliable for the cluster manually terminated or preempted on the cloud.

To reproduce the issue:
1. `sky launch -c min --cloud gcp -i 120`
2. terminate the cluster on the console manually
3. `python -c 'import sky; sky.launch(sky.Task(), cluster_name="min")'`

The following error will occur:
```
3721, in _check_existing_cluster
    cloud = to_provision.cloud
AttributeError: 'NoneType' object has no attribute 'cloud'
```

### PR DESCRIPTION
**Pull request by Michaelvll**
Fixes #2129 

We need to refactor the code in `execution` and the `_check_existing_cluster`, as the optimization and the provision are actually quite coupled, as raised by @WoosukKwon before. This PR is a fix that will unblock the issue #2129, but not the coupled issue. In the long term, we need to rearchitect the code here.

Tested (run the relevant ones):

- [x] Code formatting: `bash format.sh`
- [x] Any manual or new tests for this PR (please specify below)
  - [x] Reproducible script in #2129
- [ ] All smoke tests: `pytest tests/test_smoke.py` 
- [ ] Relevant individual smoke tests: `pytest tests/test_smoke.py::test_fill_in_the_name` 
- [ ] Backward compatibility tests: `bash tests/backward_comaptibility_tests.sh`

### CODE REVIEW

File: sky/backends/cloud_vm_ray_backend.py [hunk 1]
```diff
@@ -3701,6 +3704,20 @@ def _check_existing_cluster(
                       if resources.cloud is not None else clouds.Cloud)
         task_cloud.check_cluster_name_is_valid(cluster_name)
 
+        if to_provision is None:
+            logger.info(
+                f'The cluster {cluster_name!r} was autodowned or manually '
+                'terminated on the cloud console. Using the original resources '
+                'to provision a new cluster.')
```
Comment (WoosukKwon) [inline sky/backends/cloud_vm_ray_backend.py:3711]: I think the meaning of "original" here is a bit unclear. Can we change this like following?
```suggestion
                'terminated on the cloud console. Using the same resources as the previously terminated one'
                'to provision a new cluster.')
```

File: sky/backends/cloud_vm_ray_backend.py [hunk 2]
```diff
@@ -3701,6 +3704,20 @@ def _check_existing_cluster(
...
+            # The cluster is recently terminated either by autostop or manually
+            # terminated on the cloud. We should use the original resources to
```
Comment (WoosukKwon) [inline sky/backends/cloud_vm_ray_backend.py:3713]: ditto

File: sky/backends/cloud_vm_ray_backend.py [hunk 3]
```diff
@@ -3701,6 +3704,20 @@ def _check_existing_cluster(
...
+            # provision the cluster.
+            assert isinstance(previous_handle,
+                              CloudVmRayResourceHandle), (previous_handle,
+                                                          cluster_name)
```
Comment (WoosukKwon) [inline sky/backends/cloud_vm_ray_backend.py:3717]: Style nit: The code looks a bit ugly tbh.
```suggestion
            is_cloud_vm_handle = isinstance(previous_handle,
                                                                      CloudVmRayResourceHandle)
            assert is_cloud_vm_handle, (previous_handle, cluster_name)
```

**Review by WoosukKwon — APPROVED** [review]
LGTM. Thanks for fixing the bug. Left small comments.

File: sky/backends/cloud_vm_ray_backend.py (hunk 3 above)
Comment (Michaelvll) [reply sky/backends/cloud_vm_ray_backend.py:3717]: Unfortunately, mypy is not clear enough to have this two-line assertion. I changed the output to be a string, so as to make it look a bit better. PTAL. : )

**Pull request merged**